snuba/utils/streams/kafka.py

```diff
@@ -637,3 +645,69 @@ def close(self, timeout: Optional[float] = None) -> None:
         messages: int = self.__producer.flush(*[timeout] if timeout is not None else [])
         if messages > 0:
             raise TimeoutError(f"{messages} commit log messages pending delivery")
+
+
+class KafkaProducer(Producer[TPayload]):
+    def __init__(self, configuration, codec: Codec[KafkaPayload, TPayload]) -> None:
+        self.__configuration = configuration
+        self.__codec = codec
+
+        self.__producer = ConfluentProducer(configuration)
+        self.__shutdown_requested = Event()
+
+        self.__result = execute(self.__worker)
+
+    def __worker(self) -> None:
+        while not self.__shutdown_requested.is_set():
+            self.__producer.poll(0.1)
+        self.__producer.flush()
+
+    def __delivery_callback(
+        self,
+        future: Future[MessageDetails],
+        error: KafkaError,
+        message: ConfluentMessage,
+    ) -> None:
+        if error is not None:
+            future.set_exception(TransportError(error))
+        else:
+            future.set_result(
+                MessageDetails(
+                    Partition(Topic(message.topic()), message.partition()),
+                    message.offset(),
+                )
+            )
+
+    def produce(
+        self, destination: Union[Topic, Partition], payload: TPayload
+    ) -> Future[MessageDetails]:
+        if self.__shutdown_requested.is_set():
+            raise RuntimeError("producer has been closed")
```

Comment:
This doesn't check if the worker thread itself has exited/crashed for any reason without an exit request. It seems like it probably should, but I'm not sure what we should do if that happens. Just letting it crash is probably the safest option?